vpr/src/place/initial_placement.cpp

```diff
@@ -549,47 +547,188 @@ static std::vector<ClusterBlockId> find_centroid_loc(const t_pl_macro& pl_macro,
-        //       below / to the left due to tiny numerical changes (this
-        //       pretends that each atom is in the center of the tile).
-        centroid.x = std::floor(acc_x / acc_weight);
-        centroid.y = std::floor(acc_y / acc_weight);
-        centroid.layer = std::floor(acc_layer / acc_weight);
-        centroid.sub_tile = std::floor(acc_sub_tile / acc_weight);
-
-        // TODO: Make this a debug print.
-        // VTR_LOG("\n\t(%d, %d, %d, %d)\n", centroid.x, centroid.y, centroid.layer, centroid.sub_tile);
+        centroid /= acc_weight;
+    }
+    return centroid;
+}
+
+/**
+ * @brief Find the nearest compatible location for the given macro as close to
+ *        the src_flat_loc as possible.
+ *
+ * This method uses a BFS to find the closest legal location for the macro.
+ *
+ *  @param src_flat_loc
+ *          The start location of the BFS. This is given as a flat placement to
+ *          allow the search to trade-off different location options. For example,
+ *          if src_loc was (1.6, 1.5), this tells the search that the cluster
+ *          would prefer to be at tile (1, 1), but if it cannot go there and
+ *          it had to go to one of the neighbors, it would prefer to be on the
+ *          right.
+ *  @param block_type
+ *          The logical block type of the macro.
+ *  @param macro
+ *          The macro to place in the location.
+ *  @param blk_loc_registry
+ *
+ *  @return Returns the closest legal location found. All of the dimensions will
+ *          be OPEN if a locations could not be found.
+ */
+static inline t_pl_loc find_nearest_compatible_loc(t_flat_pl_loc& src_flat_loc,
+                                                   t_logical_block_type_ptr block_type,
+                                                   const t_pl_macro& pl_macro,
+                                                   const BlkLocRegistry& blk_loc_registry) {
+    // This method performs a BFS over the compressed grid. This avoids searching
+    // locations which obviously cannot implement this macro.
+    const auto& compressed_block_grid = g_vpr_ctx.placement().compressed_block_grids[block_type->index];
+    const int num_layers = g_vpr_ctx.device().grid.get_num_layers();
+    // This method does not support 3D FPGAs yet. The search performed will only
+    // traverse the same layer as the src_loc.
+    VTR_ASSERT(num_layers == 1);
+
+    // Get the closest (approximately) compressed location to the src location.
+    // This does not need to be perfect (in fact I do not think it is), but the
+    // closer it is, the faster the BFS will find the best solution.
+    t_physical_tile_loc src_grid_loc(src_flat_loc.x, src_flat_loc.y, src_flat_loc.layer);
+    const t_physical_tile_loc compressed_src_loc = compressed_block_grid.grid_loc_to_compressed_loc_approx(src_grid_loc);
+
+    // Weighted-BFS search the compressed grid for an empty compatible subtile.
+    size_t num_rows = compressed_block_grid.get_num_rows(0);
```

Comment:
I suggest setting layer = 0 and then passing layer to the get_num_rows and get_num_columns calls to make it more clear what you're doing, and to make it a bit easier to update to multi-layer later.